### ISSUE TITLE
Fix error "duplicate key value violates unique constraint"

### ISSUE THREAD

**Issue by mcrot** [ContactEngineering/topobank#667]
On creation of a new topography, this exception occured.
```
Internal Server Error: /manager/surface/957/new-topography/

IntegrityError at /manager/surface/957/new-topography/
duplicate key value violates unique constraint "manager_topography_surface_id_name_029e17aa_uniq"
DETAIL:  Key (surface_id, name)=(957, t1_5x.OPD) already exists.
``` 
Stacktrace:
```python
Traceback:

File "/opt/conda/lib/python3.7/site-packages/django/db/backends/utils.py" in _execute
  84.                 return self.cursor.execute(sql, params)

The above exception (duplicate key value violates unique constraint "manager_topography_surface_id_name_029e17aa_uniq"
DETAIL:  Key (surface_id, name)=(957, t1_5x.OPD) already exists.
) was the direct cause of the following exception:

File "/opt/conda/lib/python3.7/site-packages/django/core/handlers/exception.py" in inner
  34.             response = get_response(request)

File "/opt/conda/lib/python3.7/site-packages/django/core/handlers/base.py" in _get_response
  115.                 response = self.process_exception_by_middleware(e, request)

File "/opt/conda/lib/python3.7/site-packages/django/core/handlers/base.py" in _get_response
  113.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/opt/conda/lib/python3.7/contextlib.py" in inner
  74.                 return func(*args, **kwds)

File "/opt/conda/lib/python3.7/site-packages/django/contrib/auth/decorators.py" in _wrapped_view
  21.                 return view_func(request, *args, **kwargs)

File "/opt/conda/lib/python3.7/site-packages/django/views/generic/base.py" in view
  71.             return self.dispatch(request, *args, **kwargs)

File "/opt/conda/lib/python3.7/site-packages/django/contrib/auth/mixins.py" in dispatch
  109.         return super().dispatch(request, *args, **kwargs)

File "/opt/conda/lib/python3.7/site-packages/formtools/wizard/views.py" in dispatch
  244.         response = super().dispatch(request, *args, **kwargs)

File "/opt/conda/lib/python3.7/site-packages/django/views/generic/base.py" in dispatch
  97.         return handler(request, *args, **kwargs)

File "/opt/conda/lib/python3.7/site-packages/formtools/wizard/views.py" in post
  302.                 return self.render_done(form, **kwargs)

File "/opt/conda/lib/python3.7/site-packages/formtools/wizard/views.py" in render_done
  358.         done_response = self.done(final_forms.values(), form_dict=final_forms, **kwargs)

File "/app/topobank/manager/views.py" in done
  412.         instance.save()

File "/opt/conda/lib/python3.7/site-packages/django/db/models/base.py" in save
  744.                        force_update=force_update, update_fields=update_fields)

File "/opt/conda/lib/python3.7/site-packages/django/db/models/base.py" in save_base
  782.                 force_update, using, update_fields,

File "/opt/conda/lib/python3.7/site-packages/django/db/models/base.py" in _save_table
  873.             result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)

File "/opt/conda/lib/python3.7/site-packages/django/db/models/base.py" in _do_insert
  911.                                using=using, raw=raw)

File "/opt/conda/lib/python3.7/site-packages/django/db/models/manager.py" in manager_method
  82.                 return getattr(self.get_queryset(), name)(*args, **kwargs)

File "/opt/conda/lib/python3.7/site-packages/django/db/models/query.py" in _insert
  1186.         return query.get_compiler(using=using).execute_sql(return_id)

File "/opt/conda/lib/python3.7/site-packages/django/db/models/sql/compiler.py" in execute_sql
  1377.                 cursor.execute(sql, params)

File "/opt/conda/lib/python3.7/site-packages/django/db/backends/utils.py" in execute
  67.         return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)

File "/opt/conda/lib/python3.7/site-packages/django/db/backends/utils.py" in _execute_with_wrappers
  76.         return executor(sql, params, many, context)

File "/opt/conda/lib/python3.7/site-packages/django/db/backends/utils.py" in _execute
  84.                 return self.cursor.execute(sql, params)

File "/opt/conda/lib/python3.7/site-packages/django/db/utils.py" in __exit__
  89.                 raise dj_exc_value.with_traceback(traceback) from exc_value

File "/opt/conda/lib/python3.7/site-packages/django/db/backends/utils.py" in _execute
  84.                 return self.cursor.execute(sql, params)

Exception Type: IntegrityError at /manager/surface/957/new-topography/
Exception Value: duplicate key value violates unique constraint "manager_topography_surface_id_name_029e17aa_uniq"
DETAIL:  Key (surface_id, name)=(957, t1_5x.OPD) already exists.
[...]
```
Not sure yet whether this has an impact on UI level.
Maybe this can be fixed by recreating the index..to be checked.

**Comment by mcrot** [comment]
Maybe this will help: https://dba.stackexchange.com/questions/193568/how-to-fix-all-duplicate-key-value-violates-unique-constraint

**Comment by mcrot** [comment]
This does not happen on every measurement upload. I've just tested it without having problems.

**Comment by mcrot** [comment]
Also using the same measurement name in the same surface only generates a validation error, which is the expected behavior:
![grafik](https://user-images.githubusercontent.com/9113541/119966952-1e62be80-bfac-11eb-8d66-6b1255df3d3e.png)

**Comment by mcrot** [comment]
The error message which corresponds to the initial issue, from postgres logs:
```
postgres_1      | 2021-05-21T14:20:35.169980117Z 2021-05-21 14:20:35.169 UTC [27643] ERROR:  duplicate key value violates unique constraint "manager_topography_surfac
e_id_name_029e17aa_uniq"
postgres_1      | 2021-05-21T14:20:35.170016605Z 2021-05-21 14:20:35.169 UTC [27643] DETAIL:  Key (surface_id, name)=(957, t1_5x.OPD) already exists.
postgres_1      | 2021-05-21T14:20:35.170027288Z 2021-05-21 14:20:35.169 UTC [27643] STATEMENT:  INSERT INTO "manager_topography" ("surface_id", "name", "creator_id",
 "measurement_date", "description", "datafile", "datafile_format", "data_source", "size_editable", "size_x", "size_y", "unit_editable", "unit", "height_scale_editable
", "height_scale", "detrend_mode", "resolution_x", "resolution_y", "is_periodic", "thumbnail") VALUES (957, 't1_5x.OPD', <some user>, '2021-05-19'::date, '5x mag', '<some path>', 'opd', 0, false, 0.9554140269756317, 0.7165605202317238, false, 'mm', true, 1.0, 'curvature', 1280, 960, false, '') RETURNING "manage
r_topography"."id"
```
Changed file path name and user id here for privacy.

**Comment by mcrot** [comment]
A similar error happens for another table, when counting statistics:
```
postgres_1      | 2021-05-18T15:56:22.589000799Z 2021-05-18 15:56:22.588 UTC [15115] ERROR:  duplicate key value violates unique constraint "trackstats_statisticbyda_
date_metric_id_object_ty_69dfc987_uniq"
postgres_1      | 2021-05-18T15:56:22.589079484Z 2021-05-18 15:56:22.588 UTC [15115] DETAIL:  Key (date, metric_id, object_type_id, object_id, period)=(2021-05-18, 9,
 15, 881, 86400) already exists.
postgres_1      | 2021-05-18T15:56:22.589086906Z 2021-05-18 15:56:22.588 UTC [15115] STATEMENT:  INSERT INTO "trackstats_statisticbydateandobject" ("metric_id", "valu
e", "period", "object_type_id", "object_id", "date") VALUES (9, 1, 86400, 15, 881, '2021-05-18'::date) RETURNING "trackstats_statisticbydateandobject"."id"
```
These errors don't seem to be related to the other error, since they are not at similar times.

**Comment by mcrot** [comment]
Maybe the error is related to `INSERT` clauses with explicit ids, although I'm not yet, why this happens: https://dba.stackexchange.com/questions/60802/fixing-table-structure-to-avoid-error-duplicate-key-value-violates-unique-cons

**Comment by mcrot** [comment]
There is a management command which probably helps here: https://docs.djangoproject.com/en/2.2/ref/django-admin/#sqlsequencereset

**Comment by mcrot** [comment]
I'll install version 0.13.0 and will reset the sequences on the server before starting new in order to reduce downtime.

**Comment by mcrot** [comment]
The SQL has been applied. I'll shift this to 0.14.0 to check whether this issue happened again in production.

**Comment by mcrot** [comment]
Unfortunately, this still happens:
```python
Traceback:

File "/opt/conda/lib/python3.7/site-packages/django/db/backends/utils.py" in _execute
  84.                 return self.cursor.execute(sql, params)

The above exception (duplicate key value violates unique constraint "manager_topography_surface_id_name_029e17aa_uniq"
DETAIL:  Key (surface_id, name)=(1035, sty_A1_SM.003_300um.txt) already exists.
) was the direct cause of the following exception:

File "/opt/conda/lib/python3.7/site-packages/django/core/handlers/exception.py" in inner
  34.             response = get_response(request)

File "/opt/conda/lib/python3.7/site-packages/django/core/handlers/base.py" in _get_response
  115.                 response = self.process_exception_by_middleware(e, request)

File "/opt/conda/lib/python3.7/site-packages/django/core/handlers/base.py" in _get_response
  113.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/opt/conda/lib/python3.7/contextlib.py" in inner
  74.                 return func(*args, **kwds)

File "/opt/conda/lib/python3.7/site-packages/django/contrib/auth/decorators.py" in _wrapped_view
  21.                 return view_func(request, *args, **kwargs)

File "/opt/conda/lib/python3.7/site-packages/django/views/generic/base.py" in view
  71.             return self.dispatch(request, *args, **kwargs)

File "/opt/conda/lib/python3.7/site-packages/django/contrib/auth/mixins.py" in dispatch
  109.         return super().dispatch(request, *args, **kwargs)

File "/opt/conda/lib/python3.7/site-packages/formtools/wizard/views.py" in dispatch
  244.         response = super().dispatch(request, *args, **kwargs)

File "/opt/conda/lib/python3.7/site-packages/django/views/generic/base.py" in dispatch
  97.         return handler(request, *args, **kwargs)

File "/opt/conda/lib/python3.7/site-packages/formtools/wizard/views.py" in post
  302.                 return self.render_done(form, **kwargs)

File "/opt/conda/lib/python3.7/site-packages/formtools/wizard/views.py" in render_done
  358.         done_response = self.done(final_forms.values(), form_dict=final_forms, **kwargs)

File "/app/topobank/manager/views.py" in done
  418.         instance.save()

File "/opt/conda/lib/python3.7/site-packages/django/db/models/base.py" in save
  744.                        force_update=force_update, update_fields=update_fields)

File "/opt/conda/lib/python3.7/site-packages/django/db/models/base.py" in save_base
  782.                 force_update, using, update_fields,

File "/opt/conda/lib/python3.7/site-packages/django/db/models/base.py" in _save_table
  873.             result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)

File "/opt/conda/lib/python3.7/site-packages/django/db/models/base.py" in _do_insert
  911.                                using=using, raw=raw)

File "/opt/conda/lib/python3.7/site-packages/django/db/models/manager.py" in manager_method
  82.                 return getattr(self.get_queryset(), name)(*args, **kwargs)

File "/opt/conda/lib/python3.7/site-packages/django/db/models/query.py" in _insert
  1186.         return query.get_compiler(using=using).execute_sql(return_id)

File "/opt/conda/lib/python3.7/site-packages/django/db/models/sql/compiler.py" in execute_sql
  1377.                 cursor.execute(sql, params)

File "/opt/conda/lib/python3.7/site-packages/django/db/backends/utils.py" in execute
  67.         return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)

File "/opt/conda/lib/python3.7/site-packages/django/db/backends/utils.py" in _execute_with_wrappers
  76.         return executor(sql, params, many, context)

File "/opt/conda/lib/python3.7/site-packages/django/db/backends/utils.py" in _execute
  84.                 return self.cursor.execute(sql, params)

File "/opt/conda/lib/python3.7/site-packages/django/db/utils.py" in __exit__
  89.                 raise dj_exc_value.with_traceback(traceback) from exc_value

File "/opt/conda/lib/python3.7/site-packages/django/db/backends/utils.py" in _execute
  84.                 return self.cursor.execute(sql, params)

Exception Type: IntegrityError at /manager/surface/1035/new-topography/
Exception Value: duplicate key value violates unique constraint "manager_topography_surface_id_name_029e17aa_uniq"
DETAIL:  Key (surface_id, name)=(1035, sty_A1_SM.003_300um.txt) already exists.
```
Maybe this is related to #691 , because the same surface is involved.

**Comment by mcrot** [comment]
When upgrading to 0.16.0 in production, all data will be dumped and reimported, because the Postgres version changes.
I'll shift this to 0.17.0, let's see whether it still happens. In my local environment I don't see this.
The UI elements prevent the user from saving a second topography with the same name under the same surface.

**Comment by mcrot** [comment]
This still happens, but probably no longer in the context of new measurements.
Where it still happens is (see postgres logs)when updating statistics
```postgres_1      | 2021-11-28T01:52:16.005136882Z 2021-11-28 01:52:16.004 UTC [285979] ERROR:  duplicate key value violates unique constraint "trackstats_statisticbyda_date_metric_id_object_ty_69dfc987_uniq"
postgres_1      | 2021-11-28T01:52:16.005223490Z 2021-11-28 01:52:16.004 UTC [285979] DETAIL:  Key (date, metric_id, object_type_id, object_id, period)=(2021-11-28, 9, 15, 943, 86400) already exists.
postgres_1      | 2021-11-28T01:52:16.005404193Z 2021-11-28 01:52:16.004 UTC [285979] STATEMENT:  INSERT INTO "trackstats_statisticbydateandobject" ("metric_id", "value", "period", "object_type_id", "object_id", "date") VALUES (9, 1, 86400, 15, 943, '2021-11-28'::date) RETURNING "trackstats_statisticbydateandobject"."id"
postgres_1      | 2021-11-28T01:52:16.005692126Z 2021-11-28 01:52:16.005 UTC [285979] LOG:  could not send data to client: Broken pipe
postgres_1      | 2021-11-28T01:52:16.005732716Z 2021-11-28 01:52:16.005 UTC [285979] FATAL:  connection to client lost
postgres_1      | 2021-11-28T01:55:45.957391521Z 2021-11-28 01:55:45.954 UTC [286005] LOG:  unexpected EOF on client connection with an open transaction
postgres_1      | 2021-11-28T01:55:45.961610023Z 2021-11-28 01:55:45.961 UTC [286004] LOG:  unexpected EOF on client connection with an open transaction
postgres_1      | 2021-11-28T01:55:45.982844968Z 2021-11-28 01:55:45.982 UTC [286002] LOG:  could not receive data from client: Connection reset by peer
postgres_1      | 2021-11-28T01:55:45.982869906Z 2021-11-28 01:55:45.982 UTC [286002] LOG:  unexpected EOF on client connection with an open transaction
(END)
```

**Comment by pastewka** [comment]
Resolved with #1012
